docker-entrypoint-tcp.sh

```diff
@@ -6,6 +6,10 @@ set -e
 # * if hyphenated flag arguments (e.g., '-f', '-foo', or '--foo') were
 #   passed to the Docker command line
 if [ "$#" -eq 0 ] || [ "${1#-}" != "$1" ]; then
+    export CCACHE_DIR="/var/cache/ccache"
+    if [ -f "$CCACHE_DIR/ccache.conf" ]; then
```

Comment:
I'd prefer an explicit envvar (e.g., `[DISTCC_]CCACHE_ENABLED=1`) for enabling ccache instead of checking for the presence of an optional config file.

Reply:
That variable does exist and can be set. I was looking for a way to make it a "one switch" thing. The user mounts a dir being a cache at a known location, and that does it all.

If we take CCACHE_DIR and CCACHE_ENABLED (defaults to 1 if unset) into the script as well, we will clutter the code and make the user interface a combination of setting multiple env vars and the mount target. Plus the obligation to align the mount target with the dir.

Note that you can disable ccache inside the container already with `-e CCACHE_ENABLED=0` but why even bother if you can skip the mounting.

Reply:
Are you referring to the `CCACHE_DISABLE` and `CCACHE_NODISABLE` envvars defined by ccache?

Checking for the existence of a directory in order to toggle a feature is not a common Docker pattern I'm afraid, it will have to be an explicit envvar. From a UX perspective, the feature should be off by default and the user can toggle it with the envvar (without having to define/modify `CCACHE_DIR`).

Reply:
How about we move away from the code and talk about the envisioned user interface.

What i suggest is:
1. create a ccache on the host if it does not yet have one
2. mount that ccache to /var/cache/ccache in the container

where 1 is only needed the first time, we want things to persist after all

You seem to suggest one or two extra things:
3. specify mount target again in CCACHE_DIR (inside container value), any defaults?
4. enable it by setting another "enable" env var which defaults to "false"

Having to do 2+3+4 instead of just 2 sounds rather complicated to me. But maybe i did not understand it right, so please tell me how you envision it.

Reply:
The ideal workflow looks like this:
1. The user toggles the ccache flag (e.g., `CCACHE_ENABLED=1`) and the cache automatically persists in an anonymous Docker volume. This volume has been prepped in advance in the Dockerfile, so that no further configuration is required. For stronger persistence guarantees, the user may instead use a named Docker volume or a host directory.

This approach is the most common. See for instance the official redis image @ https://github.com/redis/docker-library-redis/blob/master/7.4/alpine/Dockerfile#L133 (or any other official Docker image that persists data without running as root). Using this approach `CCACHE_DIR` is probably not needed at all.

Reply:
Using the file is indeed not ideal, because `ccache` itself will work just fine without a config file. Plus there is `CCACHE_CONFIGPATH` which checking the file would kind of break.
So that `test -f` has to go and be replaced with something more specific to keep all the functionality of `ccache` and its env variables.

Using `CCACHE_DISABLE` or `CCACHE_NODISABLE` would not really work either. Because enabling any one of them means just setting them to any value. However anything like `0` or `false` is not allowed, only `unset` works. So we should not set any of them by default.

So i agree that we need a variable of our own which will trigger the PATH prefixing. The thing that does the magic and can likely not be done with `-e PATH=` trickery. But i would stay away from `CCACHE_ENABLE` because that is not "our" namespace and does not even exist in `ccache`. Might go for DISTCC_CCACHE_ENABLE as suggested before. But i have to admit i also do not like that too much.

I will give that volume a try. Can these be shared with the host or would it be only available in the "container world"?

Reply:
It turned out that CCACHE_DIR will have to be passed to both variant at `docker run` because the defaults will not work. So i decided to use that as "master switch" for both. We could have another one, but that feels wrong.

Reply:
Using `CCACHE_DIR` as a feature flag is no-go. An alternative to `[DISTCC]_CCACHE_ENABLED` would be to introduce dedicated image variants where ccache is enabled out-of-the-box, e.g., `docker-gentoo-distcc:tcp-ccache`.

Furthermore, it's crucial to use Docker volumes to enable this out-of-the-box experience, otherwise the user has to manually prep a cache directory on the host (this use case should only be supported as a fallback). The idea is that an anonymous volume is mounted in `CCACHE_DIR` by default, with the appropriate permissions. If the user requires stronger persistence guarantees, they can mount a named volume there, or go into the trouble of manually prepping and mounting a host directory.

Reply:
We can not have the user choose `CCACHE_DIR` and at the same time use `VOLUME` in the Dockerfile.

Anyhow i will likely give this a break. It is not going anywhere with the time you take to review like 4 lines of code, that by the looks of it you never tried.

Reply:
> We can not have the user choose `CCACHE_DIR` and at the same time use `VOLUME` in the Dockerfile.

We are not really interested in the user overriding `CCACHE_DIR` (setting a default would suffice), but for the sake of completeness, you can do it as follows:
```dockerfile
ENV CCACHE_DIR=/var/cache/ccache
RUN prep_ccache_dir
VOLUME $CCACHE_DIR
```

And should the user override `CCACHE_DIR` at runtime, you can handle it in the entrypoint script.

> Anyhow i will likely give this a break. It is not going anywhere with the time you take to review like 4 lines of code, that by the looks of it you never tried.

That's unfortunate; I thought I made it clear in the beginning that the review process would take a while.

Reply:
Ok i will give it one more try. The plan would be to:

- set a static `CCACHE_DIR` as in the first draft
- possibly catch a custom `CCACHE_DIR` as invalid in entrypoints
- make that `CCACHE_DIR` a `VOLUME`
- likely keep the `VOLUME` empty since any configuration is optional
- use `DISTCC_CCACHE_ENABLE` as master-switch for entrypoints

Reply:
>     * possibly catch a custom `CCACHE_DIR` as invalid in entrypoints
I don't see much value in overriding `CCACHE_DIR` at runtime, so you can safely ignore this use case for now. I would just amend my snippet above with the following, so that `CCACHE_DIR` is customizable at build time, should someone wish to roll their own images:
```dockerfile
ARG CCACHE_DIR=/var/cache/ccache
ENV CCACHE_DIR=$CCACHE_DIR
...
```
>     * use `DISTCC_CCACHE_ENABLE` as master-switch for entrypoints
Using dedicated image variants is also an acceptable solution and might be easier to reason about and implement. From a UX perspective both of them are equivalent to me since neither approach requires further interaction.

Reply:
Ok done. Went for the enable switch instead of another image.